### PR TITLE
chore: #1155 improve QR reader parser

### PR DESCRIPTION
**Pull request by prybalko**
## Purpose

improve readablilty and relibility of QR reader parser

## Parser specs
https://paritytech.github.io/parity-signer/development/UOS.html#raptorq-multipart-payload

## Scope

- re-write paser using `nom`
- add fuzzing (it takes hours to run, so I didn't put it in CI)
- fix documentation

## Screenshots

| Before |  After |
|-|-|
|image|image
|

### CODE REVIEW

File: rust/qr_reader_phone/src/parser.rs [hunk 1]
```diff
@@ -1,9 +1,67 @@
...
+        //
+        // Thus, if there are no additional packets, probability is ~ 0.99609.
+        // If one additional packet is added, it is ~ 0.99998.
+        // It was decided to add one additional packet in the printed estimate, so that
+        // the user expectations are lower.
+        self.size / (self.payload.len() as u32) + 1
+    }
+}
+
+impl TryFrom<&Vec<u8>> for RaptorqFrame {
```
Comment (montekki) [inline rust/qr_reader_phone/src/parser.rs:33]: `impl TryFrom<&[u8]>` ?

File: rust/qr_reader_phone/src/parser.rs [hunk 2]
```diff
@@ -1,9 +1,67 @@
...
+    }
+}
+
+pub(crate) struct LegacyFrame {
+    pub(crate) total: u16,
+    pub(crate) index: u16,
+    pub(crate) data: Vec<u8>,
+}
+
+impl TryFrom<&Vec<u8>> for LegacyFrame {
```
Comment (montekki) [inline rust/qr_reader_phone/src/parser.rs:52]: and here

File: rust/qr_reader_phone/src/parser.rs (hunk 1 above)
Comment (prybalko) [reply rust/qr_reader_phone/src/parser.rs:33]: I get
```
 the trait `From<&Vec<u8>>` is not implemented for `RaptorqFrame`
```

Comment (montekki) [reply rust/qr_reader_phone/src/parser.rs:33]: Yes everywhere you call `try_from(&payload)` you would have to call `try_from(payload.as_ref())` and it should work

File: rust/qr_reader_phone/fuzz/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "qr_reader_phone-fuzz"
+version = "0.0.0"
+authors = ["Automatically generated"]
+publish = false
+edition = "2018"
```
Comment (montekki) [inline rust/qr_reader_phone/fuzz/Cargo.toml:6]: The regular stuff like everywhere else should be used

File: rust/qr_reader_phone/src/parser.rs (hunk 1 above)
Comment (prybalko) [reply rust/qr_reader_phone/src/parser.rs:33]: Thanks! But is there a reason to prefer one over the other? Or it's more a matter of taste?

Comment (montekki) [reply rust/qr_reader_phone/src/parser.rs]: No actually it is quite useful. When you choose `Vec<u8>` on a type-level you are binding yourself to work only with this single type. When you choose `&[u8]` you are "saying" something like "i can work with any type that can act as a contiguous piece of memory containing bytes". For idiomatic examples you can look at how the `std::io` traits are composed. For example check out the [`std::io::Write`](https://doc.rust-lang.org/std/io/trait.Write.html) trait.

Comment (prybalko) [reply rust/qr_reader_phone/src/parser.rs]: Awesome, thanks!

**Review by montekki — APPROVED** [review]

**Pull request merged**